services/api/src/models/audit-log.js

```diff
@@ -0,0 +1,75 @@
+const mongoose = require('mongoose');
+const { createSchema } = require('../utils/schema');
+const { get, pick, isEmpty, intersection } = require('lodash');
+
+const definition = require('./definitions/audit-log.json');
+
+const schema = createSchema(definition.attributes);
+
+schema.statics.getContextFields = function (ctx) {
+  return {
+    user: ctx.state.authUser?.id,
+    requestMethod: ctx.request.method,
+    requestUrl: ctx.request.url,
+    routeNormalizedPath: ctx.routerPath,
+    routePrefix: ctx.router.opts.prefix,
+  };
+};
+
+schema.statics.getObjectFields = function getObjectFields(object, fields = [], isNew = false) {
+  const isMongooseDoc = object instanceof mongoose.Model;
+  if (!isMongooseDoc) throw Error('AuditLog.getObjectFields only works with mongoose documents');
+
+  const objectFields = {
+    objectId: object.id,
+    objectType: object.constructor.modelName,
+  };
+
+  if (fields.length) {
+    if (isNew) {
+      objectFields.objectAfter = pick(object.toObject({ depopulate: true }), fields);
+    } else {
+      const { auditOriginal, auditPathsModified } = object.$locals;
+      // the mongoose's Document.directModifiedPaths() returns falsely the objectId modified (quite possible due to autopopulate)
+      // this corrects any issues that happens when comparing with objectIds
+      const pathsModified = intersection(auditPathsModified, fields).filter((field) => {
+        if (!object.get(field).equals) return true;
+        return !object.get(field).equals(get(auditOriginal, field));
+      });
+
+      const after = pick(pick(object.toObject({ depopulate: true }), pathsModified), fields);
```

Comment:
is this the different from `pick(object, [...pathsModified, ...fields])`?
Also I'm wondering if you ever would actually use `fields` ... the audit trail almost by definition you don't know what you're interested in up front, seems like would always just be interested in all changed fields?

Reply:
`is this the different from pick(object, [...pathsModified, ...fields])?`
Yes. So the logic i am trying todo here is to ensure only get fields that changed. 
But I could possible do an _.intersect with pathsModified and fields to make that a bit cleaner

This is to my understanding one of the requirements, not to track everything but only track specify fields.
cc @dominiek

Reply:
well it looks like it's getting fields that changed but also specific ones you can pass in no?

Reply:
Should be addressed now